File: src/action.cc
#include <iostream>
#include <future>
#include <utility>

#include "action.h"
#include "system.h"

using namespace mavjs;

class ArmWorker : public Napi::AsyncWorker {
    public:
        ArmWorker(Napi::Env &env, mavsdk::Action* action, Napi::Promise::Deferred& deferred)
        : Napi::AsyncWorker(env), action(action), deferred(deferred) {}

        ~ArmWorker() {}
    
    void Execute() {
        auto prom = std::make_shared<std::promise<mavsdk::Action::Result>>();
        auto future_result = prom->get_future();
        action->arm_async(
              [prom](mavsdk::Action::Result result) { prom->set_value(result); });

        result = future_result.get();
    }

    void OnOK() {
        deferred.Resolve(Napi::String::New(Env(), mavsdk::Action::result_str(result)));
    }

    private:     
        mavsdk::Action* action;
        Napi::Promise::Deferred deferred;
        
        mavsdk::Action::Result result;
};

class DisarmWorker : public Napi::AsyncWorker {
    public:
        DisarmWorker(Napi::Env &env, mavsdk::Action* action, Napi::Promise::Deferred& deferred)
        : Napi::AsyncWorker(env), action(action), deferred(deferred) {}

        ~DisarmWorker() {}
    
    void Execute() {
        auto prom = std::make_shared<std::promise<mavsdk::Action::Result>>();
        auto future_result = prom->get_future();
        action->disarm_async(
              [prom](mavsdk::Action::Result result) { prom->set_value(result); });

        result = future_result.get();
    }

    void OnOK() {
        deferred.Resolve(Napi::String::New(Env(), mavsdk::Action::result_str(result)));
    }

    private:
        mavsdk::Action* action;
        Napi::Promise::Deferred deferred;
        
        mavsdk::Action::Result result;
};

class KillWorker : public Napi::AsyncWorker {
    public:
        KillWorker(Napi::Env &env, mavsdk::Action* action, Napi::Promise::Deferred& deferred)
        : Napi::AsyncWorker(env), action(action), deferred(deferred) {}

        ~KillWorker() {}
    
    void Execute() {
        auto prom = std::make_shared<std::promise<mavsdk::Action::Result>>();
        auto future_result = prom->get_future();
        action->kill_async(
              [prom](mavsdk::Action::Result result) { prom->set_value(result); });

        result = future_result.get();
    }

    void OnOK() {
        deferred.Resolve(Napi::String::New(Env(), mavsdk::Action::result_str(result)));
    }

    private:
        mavsdk::Action* action;
        Napi::Promise::Deferred deferred;
        
        mavsdk::Action::Result result;
};

class TakeoffWorker : public Napi::AsyncWorker {
    public:
        TakeoffWorker(Napi::Env &env, mavsdk::Action* action, Napi::Promise::Deferred& deferred)
        : Napi::AsyncWorker(env), action(action), deferred(deferred) {}

        ~TakeoffWorker() {}
    
    void Execute() {
        auto prom = std::make_shared<std::promise<mavsdk::Action::Result>>();
        auto future_result = prom->get_future();
        action->takeoff_async(
              [prom](mavsdk::Action::Result result) { prom->set_value(result); });

        result = future_result.get();
    }

    void OnOK() {
        deferred.Resolve(Napi::String::New(Env(), mavsdk::Action::result_str(result)));
    }

    private:
        mavsdk::Action* action;
        Napi::Promise::Deferred deferred;
        
        mavsdk::Action::Result result;
};

class LandWorker : public Napi::AsyncWorker {
    public:
        LandWorker(Napi::Env &env, mavsdk::Action* action, Napi::Promise::Deferred& deferred)
        : Napi::AsyncWorker(env), action(action), deferred(deferred) {}

        ~LandWorker() {}
    
    void Execute() {
        auto prom = std::make_shared<std::promise<mavsdk::Action::Result>>();
        auto future_result = prom->get_future();
        action->land_async(
              [prom](mavsdk::Action::Result result) { prom->set_value(result); });

        result = future_result.get();
    }

    void OnOK() {
        deferred.Resolve(Napi::String::New(Env(), mavsdk::Action::result_str(result)));
    }

    private:
        mavsdk::Action* action;
        Napi::Promise::Deferred deferred;
        
        mavsdk::Action::Result result;
};

class ReturnToLaunchWorker : public Napi::AsyncWorker {
    public:
        ReturnToLaunchWorker(Napi::Env &env, mavsdk::Action* action, Napi::Promise::Deferred& deferred)
        : Napi::AsyncWorker(env), action(action), deferred(deferred) {}

        ~ReturnToLaunchWorker() {}
    
    void Execute() {
        auto prom = std::make_shared<std::promise<mavsdk::Action::Result>>();
        auto future_result = prom->get_future();
        action->return_to_launch_async(
              [prom](mavsdk::Action::Result result) { prom->set_value(result); });

        result = future_result.get();
    }

    void OnOK() {
        deferred.Resolve(Napi::String::New(Env(), mavsdk::Action::result_str(result)));
    }

    private:
        mavsdk::Action* action;
        Napi::Promise::Deferred deferred;
        
        mavsdk::Action::Result result;
};

class TransitionToFixedwingWorker : public Napi::AsyncWorker {
    public:
        TransitionToFixedwingWorker(Napi::Env &env, mavsdk::Action* action, Napi::Promise::Deferred& deferred)
        : Napi::AsyncWorker(env), action(action), deferred(deferred) {}

        ~TransitionToFixedwingWorker() {}
    
    void Execute() {
        auto prom = std::make_shared<std::promise<mavsdk::Action::Result>>();
        auto future_result = prom->get_future();
        action->transition_to_fixedwing_async(
              [prom](mavsdk::Action::Result result) { prom->set_value(result); });

        result = future_result.get();
    }

    void OnOK() {
        deferred.Resolve(Napi::String::New(Env(), mavsdk::Action::result_str(result)));
    }

    private:
        mavsdk::Action* action;
        Napi::Promise::Deferred deferred;
        
        mavsdk::Action::Result result;
};

class TransitionToMulticopterWorker : public Napi::AsyncWorker {
    public:
        TransitionToMulticopterWorker(Napi::Env &env, mavsdk::Action* action, Napi::Promise::Deferred& deferred)
        : Napi::AsyncWorker(env), action(action), deferred(deferred) {}

        ~TransitionToMulticopterWorker() {}
    
    void Execute() {
        auto prom = std::make_shared<std::promise<mavsdk::Action::Result>>();
        auto future_result = prom->get_future();
        action->transition_to_multicopter_async(
              [prom](mavsdk::Action::Result result) { prom->set_value(result); });

        result = future_result.get();
    }

    void OnOK() {
        deferred.Resolve(Napi::String::New(Env(), mavsdk::Action::result_str(result)));
    }

    private:
        mavsdk::Action* action;
        Napi::Promise::Deferred deferred;
        
        mavsdk::Action::Result result;
};

Napi::FunctionReference Action::constructor;

Napi::Object Action::Init(Napi::Env env, Napi::Object exports) {
  Napi::HandleScope scope(env);

  Napi::Function func = DefineClass(env, "Action", {
    InstanceMethod("arm", &Action::arm),
    InstanceMethod("disarm", &Action::disarm),
    InstanceMethod("kill", &Action::kill),
    InstanceMethod("reboot", &Action::reboot),
    InstanceMethod("takeoff", &Action::takeoff),
    InstanceMethod("land", &Action::land),
    InstanceMethod("return_to_launch", &Action::return_to_launch),
    InstanceMethod("goto_location", &Action::goto_location),
    InstanceMethod("transition_to_fixedwing", &Action::transition_to_fixedwing),
    InstanceMethod("transition_to_multicopter", &Action::transition_to_multicopter),
    InstanceMethod("arm_async", &Action::arm_async),
    InstanceMethod("disarm_async", &Action::disarm_async),
    InstanceMethod("kill_async", &Action::kill_async),
    InstanceMethod("takeoff_async", &Action::takeoff_async),
    InstanceMethod("land_async", &Action::land_async),
    InstanceMethod("return_to_launch_async", &Action::return_to_launch_async),
    InstanceMethod("transition_to_fixedwing_async", &Action::transition_to_fixedwing_async),
    InstanceMethod("transition_to_multicopter_async", &Action::transition_to_multicopter_async),
    InstanceMethod("set_takeoff_altitude", &Action::set_takeoff_altitude),
    InstanceMethod("get_takeoff_altitude", &Action::get_takeoff_altitude),
    InstanceMethod("set_max_speed", &Action::set_max_speed),
    InstanceMethod("set_return_to_launch_return_altitude", &Action::set_return_to_launch_return_altitude),
    InstanceMethod("get_return_to_launch_return_altitude", &Action::get_return_to_launch_return_altitude)
  });

  constructor = Napi::Persistent(func);
  constructor.SuppressDestruct();

  exports.Set("Action", func);
  return exports;
}

Action::Action(const Napi::CallbackInfo& info) : Napi::ObjectWrap<Action>(info)  {
  Napi::Env env = info.Env();
  Napi::HandleScope scope(env);
  
  auto system = info[0].As<Napi::External<mavsdk::System>>().Data();
  this->_action = new mavsdk::Action(*system);
}

Napi::Value Action::arm(const Napi::CallbackInfo& info) {
    const mavsdk::Action::Result arm_result = this->_action->arm();
    return Napi::String::New(info.Env(), mavsdk::Action::result_str(arm_result));
}

Napi::Value Action::disarm(const Napi::CallbackInfo& info) {
    const mavsdk::Action::Result disarm_result = this->_action->disarm();
    return Napi::String::New(info.Env(), mavsdk::Action::result_str(disarm_result));
}

Napi::Value Action::kill(const Napi::CallbackInfo& info) {
  const mavsdk::Action::Result kill_result = this->_action->kill();
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(kill_result));
}

Napi::Value Action::reboot(const Napi::CallbackInfo& info) {
  const mavsdk::Action::Result reboot_result = this->_action->reboot();
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(reboot_result));
}

Napi::Value Action::takeoff(const Napi::CallbackInfo& info) {
  const mavsdk::Action::Result takeoff_result = this->_action->takeoff();
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(takeoff_result));
}

Napi::Value Action::land(const Napi::CallbackInfo& info) {
  const mavsdk::Action::Result land_result = this->_action->land();
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(land_result));
}

Napi::Value Action::return_to_launch(const Napi::CallbackInfo& info) {
  const mavsdk::Action::Result return_to_launch_result = this->_action->return_to_launch();
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(return_to_launch_result));
}

Napi::Value Action::goto_location(const Napi::CallbackInfo& info) {
  double latitude_deg = info[0].As<Napi::Number>().DoubleValue();
  double longitude_deg = info[1].As<Napi::Number>().DoubleValue();
  float altitude_amsl_m = info[2].As<Napi::Number>().FloatValue();
  float yaw_deg = info[3].As<Napi::Number>().FloatValue();
  const mavsdk::Action::Result goto_location_result = this->_action->goto_location(latitude_deg, longitude_deg, altitude_amsl_m, yaw_deg);
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(goto_location_result));
}

Napi::Value Action::transition_to_fixedwing(const Napi::CallbackInfo& info) {
  const mavsdk::Action::Result transition_to_fixedwing_result = this->_action->transition_to_fixedwing();
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(transition_to_fixedwing_result));
}

Napi::Value Action::transition_to_multicopter(const Napi::CallbackInfo& info) {
  const mavsdk::Action::Result transition_to_multicopter_result = this->_action->transition_to_multicopter();
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(transition_to_multicopter_result));
}

Napi::Value Action::arm_async(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  Napi::Promise::Deferred deferred = Napi::Promise::Deferred::New(env);

  ArmWorker* wk = new ArmWorker(env, this->_action, deferred);
  wk->Queue();
  return deferred.Promise();
}

Napi::Value Action::disarm_async(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  Napi::Promise::Deferred deferred = Napi::Promise::Deferred::New(env);

  DisarmWorker* wk = new DisarmWorker(env, this->_action, deferred);
  wk->Queue();
  return deferred.Promise();
}

Napi::Value Action::kill_async(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  Napi::Promise::Deferred deferred = Napi::Promise::Deferred::New(env);

  KillWorker* wk = new KillWorker(env, this->_action, deferred);
  wk->Queue();
  return deferred.Promise();
}

Napi::Value Action::takeoff_async(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  Napi::Promise::Deferred deferred = Napi::Promise::Deferred::New(env);

  TakeoffWorker* wk = new TakeoffWorker(env, this->_action, deferred);
  wk->Queue();
  return deferred.Promise();
}

Napi::Value Action::land_async(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  Napi::Promise::Deferred deferred = Napi::Promise::Deferred::New(env);

  LandWorker* wk = new LandWorker(env, this->_action, deferred);
  wk->Queue();
  return deferred.Promise();
}

Napi::Value Action::return_to_launch_async(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  Napi::Promise::Deferred deferred = Napi::Promise::Deferred::New(env);

  ReturnToLaunchWorker* wk = new ReturnToLaunchWorker(env, this->_action, deferred);
  wk->Queue();
  return deferred.Promise();
}

Napi::Value Action::transition_to_fixedwing_async(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  Napi::Promise::Deferred deferred = Napi::Promise::Deferred::New(env);

  TransitionToFixedwingWorker* wk = new TransitionToFixedwingWorker(env, this->_action, deferred);
  wk->Queue();
  return deferred.Promise();
}

Napi::Value Action::transition_to_multicopter_async(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  Napi::Promise::Deferred deferred = Napi::Promise::Deferred::New(env);

  TransitionToMulticopterWorker* wk = new TransitionToMulticopterWorker(env, this->_action, deferred);
  wk->Queue();
  return deferred.Promise();
}

Napi::Value Action::set_takeoff_altitude(const Napi::CallbackInfo& info) {
  float relative_altitude_m = info[0].As<Napi::Number>().FloatValue();
  const mavsdk::Action::Result set_takeoff_altitude_result = this->_action->set_takeoff_altitude(relative_altitude_m);
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(set_takeoff_altitude_result));
}

Napi::Value Action::get_takeoff_altitude(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  std::pair<mavsdk::Action::Result, float> get_takeoff_altitude_result = this->_action->get_takeoff_altitude();
  
  mavsdk::Action::Result result = get_takeoff_altitude_result.first;
  float value = get_takeoff_altitude_result.second;

  Napi::Object obj = Napi::Object::New(env);
  obj.Set(Napi::String::New(env, "result"), mavsdk::Action::result_str(result));
  obj.Set(Napi::String::New(env, "value"), value);

  return obj;
}

Napi::Value Action::set_max_speed(const Napi::CallbackInfo& info) {
  float speed_m_s = info[0].As<Napi::Number>().FloatValue();
  const mavsdk::Action::Result set_max_speed_result = this->_action->set_max_speed(speed_m_s);
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(set_max_speed_result));
}

Napi::Value Action::get_max_speed(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  std::pair<mavsdk::Action::Result, float> get_max_speed_result = this->_action->get_max_speed();
  
  mavsdk::Action::Result result = get_max_speed_result.first;
  float value = get_max_speed_result.second;

  Napi::Object obj = Napi::Object::New(env);
  obj.Set(Napi::String::New(env, "result"), mavsdk::Action::result_str(result));
  obj.Set(Napi::String::New(env, "value"), value);

  return obj;
}

Napi::Value Action::set_return_to_launch_return_altitude(const Napi::CallbackInfo& info) {
  float relative_altitude_m = info[0].As<Napi::Number>().FloatValue();
  const mavsdk::Action::Result set_return_to_launch_return_altitude_result = this->_action->set_return_to_launch_return_altitude(relative_altitude_m);
  return Napi::String::New(info.Env(), mavsdk::Action::result_str(set_return_to_launch_return_altitude_result));
}

Napi::Value Action::get_return_to_launch_return_altitude(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  std::pair<mavsdk::Action::Result, float> get_return_to_launch_return_altitude_result = this->_action->get_return_to_launch_return_altitude();
  
  mavsdk::Action::Result result = get_return_to_launch_return_altitude_result.first;
  float value = get_return_to_launch_return_altitude_result.second;

  Napi::Object obj = Napi::Object::New(env);
  obj.Set(Napi::String::New(env, "result"), mavsdk::Action::result_str(result));
  obj.Set(Napi::String::New(env, "value"), value);

  return obj;
}

void Action::dispose() {
  delete this->_action;
}

File: lib/action.ts
/**
 * The Action class enables simple actions for a drone such as arming, taking off, and landing.
 * 
 * Synchronous and asynchronous variants of the action methods are supplied.
 * 
 * The action methods send their associated MAVLink commands to the vehicle and complete (return synchronously or callback asynchronously) with an [[Action.Result]] value indicating whether the vehicle has accepted or rejected the command, or that there has been some error. If the command is accepted, the vehicle will then start to perform the associated action.
 * 
 * The plugin is typically created as shown below: ```let action = system.action()```
 */
export class Action {
  
  /**
   * Send command to arm the drone (synchronous).
   * 
   * Note: Arming a drone normally causes motors to spin at idle. Before arming take all safety precautions and stand clear of the drone!
   * @returns Result of request.
   */
  arm(): Action.Result {
    return this.arm();
  }

  /**
   * Send command to disarm the drone (synchronous).
   * 
   * This will disarm a drone that considers itself landed. If flying, the drone should reject the disarm command. Disarming means that all motors will stop.
   * @returns Result of request.
   */
  disarm(): Action.Result {
    return this.disarm();
  }

  /**
   * Send command to kill the drone (synchronous).
   * 
   * This will disarm a drone irrespective of whether it is landed or flying. Note that the drone will fall out of the sky if this command is used while flying.
   * @returns Result of request.
   */
  kill(): Action.Result {
    return this.kill();
  }

  /**
   * Send command to reboot the drone components.
   * 
   * This will reboot the autopilot, onboard computer, camera and gimbal.
   * @returns Result of request.
   */
  reboot(): Action.Result {
    return this.reboot();
  }

  /**
   * Send command to take off and hover (synchronous).
   * 
   * This switches the drone into position control mode and commands it to take off and hover at the takeoff altitude (set using [[set_takeoff_altitude()]]).
   * 
   * Note: that the vehicle must be armed before it can take off.
   * @returns Result of request.
   */
  takeoff(): Action.Result {
    return this.takeoff();
  }

  /**
   * Send command to land at the current position (synchronous).
   * 
   * This switches the drone to [Land mode](https://docs.px4.io/v1.9.0/en/flight_modes/land.html).
   * @returns Result of request.
   */
  land(): Action.Result {
    return this.land();
  }

  /**
   * Send command to return to the launch (takeoff) position and land (asynchronous).
   * 
   * This switches the drone into [RTL mode](https://docs.px4.io/en/flight_modes/rtl.html) which generally means it will rise up to a certain altitude to clear any obstacles before heading back to the launch (takeoff) position and land there.
   * @returns Result of request.
   */
  return_to_launch(): Action.Result {
    return this.return_to_launch();
  }

  /**
   * Send command to move the vehicle to a specific global position.
   * 
   * The latitude and longitude are given in degrees (WGS84 frame) and the altitude in meters AMSL (above mean sea level).
   * @param latitude_deg Latitude in degrees.
   * @param longitude_deg Longitude in degrees.
   * @param altitude_amsl_m Altitude AMSL in meters.
   * @param yaw_deg Yaw angle in degrees (Frame is NED, 0 is North, positive is clockwise).
   * @returns Result of request.
   */
  goto_location(latitude_deg: number, longitude_deg: number, altitude_amsl_m: number, yaw_deg: number): Action.Result {
    return this.goto_location(latitude_deg, longitude_deg, altitude_amsl_m, yaw_deg);
  }

  /**
   * Send command to transition the drone to fixedwing.
   * 
   * The associated action will only be executed for VTOL vehicles (on other vehicle types the command will fail with an Result). The command will succeed if called when the vehicle is already in fixedwing mode.
   * @returns Result of request.
   */
  transition_to_fixedwing(): Action.Result {
    return this.transition_to_fixedwing();
  }

  /**
   * Send command to transition the drone to multicopter.
   * 
   * The associated action will only be executed for VTOL vehicles (on other vehicle types the command will fail with an Result). The command will succeed if called when the vehicle is already in multicopter mode.
   * @returns Result of request.
   */
  transition_to_multicopter(): Action.Result {
    return this.transition_to_multicopter();
  }

  /**
   * Send command to arm the drone (asynchronous).
   * Note that arming a drone normally means that the motors will spin at idle. Therefore, before arming take all safety precautions and stand clear of the drone.
   */
  arm_async(): Promise<Action.Result> {
    return this.arm_async();
  }
  
  /**
   * Send command to disarm the drone (asynchronous).
   * 
   * This will disarm a drone that considers itself landed. If flying, the drone should reject the disarm command. Disarming means that all motors will stop.
   */
  disarm_async(): Promise<Action.Result> {
    return this.disarm_async();
  }

  /**
   * Send command to kill the drone (asynchronous).
   * 
   * This will disarm a drone irrespective of whether it is landed or flying. Note that the drone will fall out of the sky if you use this command while flying.
   */
  kill_async(): Promise<Action.Result> {
    return this.kill_async();
  }

  /**
   * Send command to take off and hover (asynchronous).
   * 
   * This switches the drone into position control mode and commands it to take off and hover at the takeoff altitude set using [[set_takeoff_altitude()]].
   * 
   * Note that the vehicle must be armed before it can take off.
   */
  takeoff_async(): Promise<Action.Result> {
    return this.takeoff_async();
  }

  /**
   * Send command to land at the current position (asynchronous).
   * 
   * This switches the drone to [Land mode](https://docs.px4.io/v1.9.0/en/flight_modes/land.html).
   */
  land_async(): Promise<Action.Result> {
    return this.land_async();
  }

  /**
   * Send command to return to the launch (takeoff) position and land (asynchronous).
   * 
   * This switches the drone into [RTL mode](https://docs.px4.io/en/flight_modes/rtl.html) which generally means it will rise up to a certain altitude to clear any obstacles before heading back to the launch (takeoff) position and land there.
   */
  return_to_launch_async(): Promise<Action.Result> {
    return this.return_to_launch_async();
  }

  /**
   * Send command to transition the drone to fixedwing (asynchronous).
   * 
   * The associated action will only be executed for VTOL vehicles (on other vehicle types the command will fail with an Result). The command will succeed if called when the vehicle is already in fixedwing mode.
   */
  transition_to_fixedwing_async(): Promise<Action.Result> {
    return this.transition_to_fixedwing_async();
  }

  /**
   * Send command to transition the drone to multicopter (asynchronous).
   * 
   * The associated action will only be executed for VTOL vehicles (on other vehicle types the command will fail with an Result). The command will succeed if called when the vehicle is already in multicopter mode.
   */
  transition_to_multicopter_async(): Promise<Action.Result> {
    return this.transition_to_multicopter_async();
  }
  
  /**
   * Set takeoff altitude above ground.
   * @param relative_altitude_m 
   * @returns Result of request.
   */
  set_takeoff_altitude(relative_altitude_m: number): Action.Result {
    return this.set_takeoff_altitude(relative_altitude_m);
  }

  /**
   * Get the takeoff altitude.
   * @returns A pair containing the result of request and if successful, the takeoff altitude relative to ground/takeoff location, in meters.
   */
  get_takeoff_altitude(): Action.pair_result {
    return this.get_takeoff_altitude();
  }

  /**
   * Set vehicle maximum speed.
   * @param speed_m_s Maximum speed in metres/second.
   * @returns Result of request.
   */
  set_max_speed(speed_m_s: number): Action.Result {
    return this.set_max_speed(speed_m_s);
  }
  
  /**
   * Get the vehicle maximum speed.
   * @returns A pair containing the result of the request and if successful, the maximum speed in metres/second.
   */
  get_max_speed(): Action.pair_result {
    return this.get_max_speed();
  }

  /**
   * Set the return to launch minimum return altitude.
   * 
   * When return to launch is initiated, the vehicle climbs to the return altitude if it is lower and stays at the current altitude if higher than the return altitude. Then it returns to the home location and lands there.
   * @param relative_altitude_m Return altitude relative to takeoff location, in meters.
   * @returns Result of request.
   */
  set_return_to_launch_return_altitude(relative_altitude_m: number): Action.Result {
    return this.set_return_to_launch_return_altitude(relative_altitude_m);
  }

  /**
   * Get the return to launch minimum return altitude.
   * @returns A pair containing the result of the request and if successful, the return altitude relative to takeoff location, in meters.
   */
  get_return_to_launch_return_altitude(): Action.pair_result {
    return this.get_return_to_launch_return_altitude();
  }
}


export namespace Action {

    export enum Result {
      UNKNOWN = "Unknown",
      SUCCESS = "Success",
      NO_SYSTEM = "No system",
      CONNECTION_ERROR = "Connection error",
      BUSY = "Busy",
      COMMAND_DENIED = "Command denied",
      COMMAND_DENIED_LANDED_STATE_UNKNOWN = "Command denied landed_state unknown",
      COMMAND_DENIED_NOT_LANDED = "Command denied not landed",
      TIMEOUT = "Timeout",
      VTOL_TRANSITION_SUPPORT_UNKNOWN = "VTOL transistion support unknown",
      NO_VTOL_TRANSITION_SUPPORT = "No VTOL transition support",
      PARAMETER_ERROR = "Parameter error"
    }
    
    export interface pair_result{
      result: string,
      value: number | string
    }
}
